src/mysql/MySqlConnection.cc

```diff
@@ -155,7 +155,7 @@ bool MySqlConnection::queryUnbuffered(std::string const& query) {
 int MySqlConnection::cancel() {
     std::lock_guard<std::mutex> lock(_interruptMutex);
     int rc;
-    if (!_isExecuting || _interrupted) {
+    if (_interrupted) {
```

Comment:
Is this the bug with query cancelation you found?

Reply:
Formally, I wouldn't really call this a bug. The code was designed on the assumption that the method could be called only for queries that are still in the "executing" state before the blocking MySQL's function `mysql_real_query` returns to a caller the query completion status. When `mysql_real_query` is finished, MySQL has a result set (or an iterator of such result set) that still needs to be extracted. And before the result retrieval is finished, MySQL has to maintain the state of the query and keep the relevant (substantial) temporary resources that are needed for that. We also know that (depending on the size of a result set and the back pressure from Qserv), the result extraction phase may be (and indeed, it is) much longer than the query execution phase. This is especially true when Qserv worker is streaming results back to `Czar` over the SSI protocol. And things are getting significantly worse if the result set is (very) large. So, my fix solves this problem by terminating queries at both the "execution" and  "result extraction" phases. In the latter case `wbd::QueryRunner` would just fail to retrieve the next batch of rows from the connection. This isn't a problem since the task and the MySQL connection would be already marked as canceled.